server/datastore/cached_mysql/cached_mysql.go

```diff
@@ -2,94 +2,56 @@ package cached_mysql
 
 import (
 	"context"
-	"encoding/json"
-	"fmt"
 	"time"
 
-	"github.com/fleetdm/fleet/v4/server/datastore/redis"
 	"github.com/fleetdm/fleet/v4/server/fleet"
-	redigo "github.com/gomodule/redigo/redis"
+	"github.com/patrickmn/go-cache"
 )
 
 type cachedMysql struct {
 	fleet.Datastore
 
-	redisPool fleet.RedisPool
+	c *cache.Cache
 }
 
 const (
-	CacheKeyAppConfig = "cache:AppConfig"
+	AppConfigKey               = "AppConfig"
+	DefaultAppConfigExpiration = 1 * time.Second
 )
 
-func New(ds fleet.Datastore, redisPool fleet.RedisPool) fleet.Datastore {
+func New(ds fleet.Datastore) fleet.Datastore {
 	return &cachedMysql{
 		Datastore: ds,
-		redisPool: redisPool,
+		c:         cache.New(5*time.Minute, 10*time.Minute),
 	}
 }
 
-func (ds *cachedMysql) storeInRedis(key string, v interface{}) error {
-	conn := redis.ConfigureDoer(ds.redisPool, ds.redisPool.Get())
-	defer conn.Close()
-
-	b, err := json.Marshal(v)
-	if err != nil {
-		return fmt.Errorf("marshaling object to cache in redis: %w", err)
-	}
-
-	if _, err := conn.Do("SET", key, b, "EX", (24 * time.Hour).Seconds()); err != nil {
-		return fmt.Errorf("caching object in redis: %w", err)
-	}
-
-	return nil
-}
-
-func (ds *cachedMysql) getFromRedis(key string, v interface{}) error {
-	conn := redis.ReadOnlyConn(ds.redisPool,
-		redis.ConfigureDoer(ds.redisPool, ds.redisPool.Get()))
-	defer conn.Close()
-
-	data, err := redigo.Bytes(conn.Do("GET", key))
-	if err != nil {
-		return fmt.Errorf("getting value from cache: %w", err)
-	}
-
-	err = json.Unmarshal(data, v)
-	if err != nil {
-		return fmt.Errorf("unmarshaling object from cache: %w", err)
-	}
-
-	return nil
-}
-
 func (ds *cachedMysql) NewAppConfig(ctx context.Context, info *fleet.AppConfig) (*fleet.AppConfig, error) {
 	ac, err := ds.Datastore.NewAppConfig(ctx, info)
 	if err != nil {
 		return nil, err
 	}
 
-	err = ds.storeInRedis(CacheKeyAppConfig, ac)
+	ds.c.Set(AppConfigKey, ac, DefaultAppConfigExpiration)
 
-	return ac, err
+	return ac, nil
 }
 
 func (ds *cachedMysql) AppConfig(ctx context.Context) (*fleet.AppConfig, error) {
-	ac := &fleet.AppConfig{}
-	ac.ApplyDefaults()
-
-	err := ds.getFromRedis(CacheKeyAppConfig, ac)
-	if err == nil {
-		return ac, nil
+	cachedAc, found := ds.c.Get(AppConfigKey)
+	if found {
+		copyAc := *cachedAc.(*fleet.AppConfig)
+		return &copyAc, nil
 	}
 
-	ac, err = ds.Datastore.AppConfig(ctx)
+	ac, err := ds.Datastore.AppConfig(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	err = ds.storeInRedis(CacheKeyAppConfig, ac)
+	ds.c.Set(AppConfigKey, ac, DefaultAppConfigExpiration)
 
-	return ac, err
+	return ac, nil
```

Comment:
You are not copying/cloning this one.

Reply:
Alternatively, we could document that the result of `AppConfig()` should not be modified, otherwise other goroutines may race.